File: scripts/import_large_osm_pbf.py
#!/usr/bin/env python3
"""
Python Script: Import large OSM PBF (.osm.pbf) directly into PostgreSQL database.
Uses `osmium` (C++ high-performance streaming parser) and psycopg2 execute_values for maximum throughput.

Requirements:
    pip install osmium psycopg2-binary

Usage:
    export DATABASE_URL="postgresql://user:password@host:port/dbname"
    python3 scripts/import_large_osm_pbf.py /path/to/vietnam-latest.osm.pbf [batch_size]

Example:
    python3 scripts/import_large_osm_pbf.py ./vietnam-latest.osm.pbf 500
"""

import sys
import os
import json
import time
import psycopg2
from psycopg2.extras import execute_values
try:
    import osmium
except ImportError:
    print("❌ Thư viện 'osmium' chưa được cài đặt. Vui lòng chạy: pip install osmium psycopg2-binary")
    sys.exit(1)

# Vietnam approximate Bounding Box filter
VN_BBOX = {
    'min_lon': 102.0,
    'max_lon': 110.5,
    'min_lat': 8.0,
    'max_lat': 24.0,
}

def get_admin_meta(tags):
    admin_level_raw = tags.get('admin_level')
    place = tags.get('place')
    building = tags.get('building')

    if admin_level_raw:
        try:
            lvl = int(admin_level_raw)
            if lvl == 2:
                return 2, 'Quốc gia (Country)', 7, 'country'
            elif lvl == 3:
                return 3, 'Vùng địa lý (Region)', 6, 'region'
            elif lvl == 4:
                return 4, 'Tỉnh / Thành phố trực thuộc TW', 5, 'province'
            elif lvl == 6:
                return 6, 'Quận / Huyện / Thị xã', 4, 'district'
            elif lvl == 8:
                return 8, 'Phường / Xã / Thị trấn', 3, 'ward'
            else:
                return lvl, f'Cấp hành chính {lvl}', 2, 'subdistrict'
        except ValueError:
            pass

    if building and building != 'no':
        return 'building', 'Tòa nhà / Công trình (Building)', 0, 'building'

    if place == 'country':
        return 2, 'Quốc gia', 7, 'country'
    elif place in ('state', 'province'):
        return 4, 'Tỉnh / Thành phố', 5, 'province'
    elif place in ('city', 'county', 'district'):
        return 6, 'Quận / Huyện', 4, 'district'
    elif place in ('town', 'village', 'suburb', 'quarter'):
        return 8, 'Phường / Xã', 3, 'ward'

    return 'poi', 'Địa điểm / Tiện ích (POI)', 0, 'poi'

def compute_bounding_box(coords):
    min_lon = min(p[0] for p in coords)
    max_lon = max(p[0] for p in coords)
    min_lat = min(p[1] for p in coords)
    max_lat = max(p[1] for p in coords)
    return min_lon, min_lat, max_lon, max_lat

def calculate_approx_area_km2(coords):
    if len(coords) < 3:
        return 0.0
    area = 0.0
    for i in range(len(coords) - 1):
        p1 = coords[i]
        p2 = coords[i + 1]
        area += (p2[0] - p1[0]) * (p2[1] + p1[1])
    return abs(area / 2.0) * 111.0 * 106.0

def flush_batch(cursor, batch, source_name):
    if not batch:
        return 0

    query = """
    INSERT INTO osm_places (
        id, osm_id, osm_type, name, name_vi, name_en,
        admin_level, admin_level_label, priority_rank, place_type,
        tags, min_lon, min_lat, max_lon, max_lat,
        center_lon, center_lat, geometry_type, geometry,
        area_approx_km2, source
    ) VALUES %s
    ON CONFLICT (id) DO UPDATE SET
        name = EXCLUDED.name,
        name_vi = EXCLUDED.name_vi,
        name_en = EXCLUDED.name_en,
        tags = EXCLUDED.tags,
        min_lon = EXCLUDED.min_lon,
        min_lat = EXCLUDED.min_lat,
        max_lon = EXCLUDED.max_lon,
        max_lat = EXCLUDED.max_lat,
        center_lon = EXCLUDED.center_lon,
        center_lat = EXCLUDED.center_lat,
        geometry_type = EXCLUDED.geometry_type,
        geometry = EXCLUDED.geometry,
        area_approx_km2 = EXCLUDED.area_approx_km2,
        source = EXCLUDED.source;
    """

    records = []
    for item in batch:
        records.append((
            item['id'],
            item['osm_id'],
            item['osm_type'],
            item['name'],
            item.get('name_vi'),
            item.get('name_en'),
            str(item['admin_level']),
            item['admin_level_label'],
            item['priority_rank'],
            item['place_type'],
            json.dumps(item['tags']),
            item['min_lon'],
            item['min_lat'],
            item['max_lon'],
            item['max_lat'],
            item['center_lon'],
            item['center_lat'],
            item['geometry_type'],
            json.dumps(item['geometry']),
            item.get('area_approx_km2'),
            source_name
        ))

    execute_values(cursor, query, records)
    count = len(batch)
    batch.clear()
    return count

class OSMHandler(osmium.SimpleHandler):
    def __init__(self, cursor, file_name, batch_size=500):
        super(OSMHandler, self).__init__()
        self.cursor = cursor
        self.file_name = file_name
        self.batch_size = batch_size
        self.batch_buffer = []
        self.total_saved = 0
        self.node_coords = {}
        self.pending_ways = {}
        self.nodes_count = 0
        self.ways_count = 0
        self.last_print = time.time()

    def node(self, n):
        self.nodes_count += 1
        if not n.location.valid():
            return
        lon = n.location.lon
        lat = n.location.lat

        if not (VN_BBOX['min_lat'] <= lat <= VN_BBOX['max_lat'] and VN_BBOX['min_lon'] <= lon <= VN_BBOX['max_lon']):
            return

        self.node_coords[n.id] = (lon, lat)

        tags = {t.k: t.v for t in n.tags}
        name = tags.get('name') or tags.get('name:vi') or tags.get('name:en')
        if name and (tags.get('place') or tags.get('amenity') or tags.get('tourism') or
                     tags.get('historic') or tags.get('leisure') or tags.get('shop')):
            admin_lvl, admin_lbl, rank, p_type = get_admin_meta(tags)
            self.batch_buffer.append({
                'id': f'node/{n.id}',
                'osm_id': n.id,
                'osm_type': 'node',
                'name': name,
                'name_vi': tags.get('name:vi', name),
                'name_en': tags.get('name:en', ''),
                'admin_level': admin_lvl,
                'admin_level_label': admin_lbl,
                'priority_rank': rank,
                'place_type': p_type,
                'tags': tags,
                'min_lon': lon - 0.001,
                'max_lon': lon + 0.001,
                'min_lat': lat - 0.001,
                'max_lat': lat + 0.001,
                'center_lon': lon,
                'center_lat': lat,
                'geometry_type': 'Point',
                'geometry': {'coordinates': [lon, lat]},
                'area_approx_km2': 0.01
            })

            if len(self.batch_buffer) >= self.batch_size:
                self.total_saved += flush_batch(self.cursor, self.batch_buffer, self.file_name)

        now = time.time()
        if now - self.last_print > 3:
            self.last_print = now
            print(f"\r⏳ Tiến trình đọc PBF: {self.nodes_count:,} nodes | Đã lưu: {self.total_saved:,}", end='', flush=True)

    def way(self, w):
        self.ways_count += 1
        tags = {t.k: t.v for t in w.tags}
        is_bldg = bool(tags.get('building') and tags.get('building') != 'no')
        is_bndry = bool(tags.get('boundary') == 'administrative' or tags.get('admin_level'))

        if (is_bldg or is_bndry) and len(w.nodes) >= 3:
            refs = [n.ref for n in w.nodes]
            coords = [self.node_coords[r] for r in refs if r in self.node_coords]
            if len(coords) < 3:
                return

            if coords[0] != coords[-1]:
                coords.append(coords[0])

            min_lon, min_lat, max_lon, max_lat = compute_bounding_box(coords)
            center_lon = (min_lon + max_lon) / 2.0
            center_lat = (min_lat + max_lat) / 2.0
            area = calculate_approx_area_km2(coords)

            admin_lvl, admin_lbl, rank, p_type = get_admin_meta(tags)
            name = tags.get('name') or tags.get('name:vi') or (f"Tòa nhà #{w.id}" if is_bldg else f"Ranh giới #{w.id}")

            self.batch_buffer.append({
                'id': f'way/{w.id}',
                'osm_id': w.id,
                'osm_type': 'way',
                'name': name,
                'name_vi': tags.get('name:vi', name),
                'name_en': tags.get('name:en', ''),
                'admin_level': admin_lvl,
                'admin_level_label': admin_lbl,
                'priority_rank': 0 if is_bldg else rank,
                'place_type': 'building' if is_bldg else p_type,
                'tags': tags,
                'min_lon': min_lon,
                'max_lon': max_lon,
                'min_lat': min_lat,
                'max_lat': max_lat,
                'center_lon': center_lon,
                'center_lat': center_lat,
                'geometry_type': 'Polygon',
                'geometry': {'coordinates': [coords]},
                'area_approx_km2': area
            })

            if len(self.batch_buffer) >= self.batch_size:
                self.total_saved += flush_batch(self.cursor, self.batch_buffer, self.file_name)

def main():
    if len(sys.argv) < 2:
        print("❌ Thiếu đường dẫn file PBF.")
        print("Cách dùng:")
        print("    python3 scripts/import_large_osm_pbf.py <duong-dan-file.osm.pbf> [batch_size]")
        sys.exit(1)

    pbf_path = os.path.abspath(sys.argv[1])
    batch_size = int(sys.argv[2]) if len(sys.argv) > 2 else 500

    if not os.path.exists(pbf_path):
        print(f"❌ Không tìm thấy file: {pbf_path}")
        sys.exit(1)

    file_size_bytes = os.path.getsize(pbf_path)
    file_size_mb = file_size_bytes / (1024 * 1024)
    file_name = os.path.basename(pbf_path)

    # Database connection
    db_url = os.getenv("DATABASE_URL")
    if not db_url:
        host = os.getenv("SQL_HOST") or os.getenv("PGHOST") or "localhost"
        user = os.getenv("SQL_USER") or os.getenv("PGUSER") or "postgres"
        password = os.getenv("SQL_PASSWORD") or os.getenv("PGPASSWORD") or ""
        dbname = os.getenv("SQL_DB_NAME") or os.getenv("PGDATABASE") or "osm_vietnam"
        port = os.getenv("SQL_PORT") or os.getenv("PGPORT") or "5432"
        db_url = f"postgresql://{user}:{password}@{host}:{port}/{dbname}"

    print("====================================================")
    print("🚀 PYTHON STREAMING OSM PBF IMPORTER (PYOSMIUM)")
    print(f"📁 File: {file_name} ({file_size_mb:.2f} MB)")
    print(f"📦 Batch Size: {batch_size}")
    print("====================================================")

    conn = psycopg2.connect(db_url)
    conn.autocommit = True
    cur = conn.cursor()

    start_time = time.time()
    handler = OSMHandler(cur, file_name, batch_size)

    # Apply osmium stream (locations=True caches node coordinates in RAM or temporary file)
    handler.apply_file(pbf_path, locations=True)

    # Flush remaining
    if handler.batch_buffer:
        handler.total_saved += flush_batch(cur, handler.batch_buffer, file_name)

    # Audit log
    cur.execute(
        "INSERT INTO import_logs (filename, places_count, file_size, status) VALUES (%s, %s, %s, %s)",
        (file_name, handler.total_saved, file_size_bytes, 'success')
    )

    cur.close()
    conn.close()

    duration = time.time() - start_time
    print(f"\n====================================================")
    print(f"✅ THÀNH CÔNG: Đã lưu {handler.total_saved:,} địa điểm từ PBF vào PostgreSQL trong {duration:.1f}s!")
    print(f"====================================================")

if __name__ == '__main__':
    main()

File: scripts/import_large_osm_pbf.ts
/**
 * Production-Grade 2-Pass OSM PBF Importer directly into PostgreSQL
 *
 * Designed to handle full-country OSM PBF files (like Vietnam 314MB with 46M nodes)
 * with low memory footprint (~500MB RAM) using coordinate bit-packing and 2-pass indexing.
 *
 * Usage:
 *   NODE_OPTIONS=--max-old-space-size=4096 npx tsx scripts/import_large_osm_pbf.ts ./vietnam-260922.osm.pbf 500
 */

import fs from 'fs';
import path from 'path';
// @ts-ignore - osm-pbf-parser has no types declaration
import parseOSM from 'osm-pbf-parser';
import { db } from '../src/db/index.ts';
import { osmPlaces, importLogs } from '../src/db/schema.ts';
import { sql } from 'drizzle-orm';
import { OSMPlace, OSMTags, GeoPolygon, GeoMultiPolygon } from '../server/types.ts';
import { computeBoundingBox } from '../server/spatialIndex.ts';
import { getAdminMeta, formatBuildingName } from '../server/pbfParser.ts';

// Bounding box filter for Vietnam
const VN_BBOX = {
  minLon: 102.0,
  maxLon: 110.5,
  minLat: 8.0,
  maxLat: 24.0,
};

// Compact coordinate packing: encode lon/lat into a single 53-bit JS number (10cm precision)
function packCoord(lon: number, lat: number): number {
  const iLat = Math.round((lat - 8.0) * 1000000);
  const iLon = Math.round((lon - 102.0) * 1000000);
  return iLat * 20000000 + iLon;
}

function unpackCoord(packed: number): [number, number] {
  const iLat = Math.floor(packed / 20000000);
  const iLon = packed % 20000000;
  return [
    Number((102.0 + iLon / 1000000).toFixed(6)),
    Number((8.0 + iLat / 1000000).toFixed(6)),
  ];
}

function calculateApproxAreaKm2(rings: number[][][]): number {
  if (!rings || rings.length === 0 || !rings[0] || rings[0].length < 3) return 0;
  const ring = rings[0];
  let area = 0;
  for (let i = 0; i < ring.length - 1; i++) {
    const p1 = ring[i];
    const p2 = ring[i + 1];
    area += (p2[0] - p1[0]) * (p2[1] + p1[1]);
  }
  return Math.abs(area / 2) * 111 * 106;
}

async function flushBatch(batch: OSMPlace[], sourceName: string): Promise<number> {
  if (batch.length === 0) return 0;

  const rows = batch.map((place) => ({
    id: place.id,
    osmId: Number(place.osmId),
    osmType: place.osmType,
    name: place.name,
    nameVi: place.nameVi ?? null,
    nameEn: place.nameEn ?? null,
    adminLevel: String(place.adminLevel),
    adminLevelLabel: place.adminLevelLabel,
    priorityRank: place.placeType === 'building' ? 0 : Math.round(Number(place.priorityRank) || 0),
    placeType: place.placeType,
    tags: place.tags || {},
    minLon: place.bbox.minLon,
    minLat: place.bbox.minLat,
    maxLon: place.bbox.maxLon,
    maxLat: place.bbox.maxLat,
    centerLon: place.center[0],
    centerLat: place.center[1],
    geometryType: place.geometryType,
    geometry: place.geometry,
    areaApproxKm2: place.areaApproxKm2 ?? null,
    source: sourceName,
  }));

  await db
    .insert(osmPlaces)
    .values(rows)
    .onConflictDoUpdate({
      target: osmPlaces.id,
      set: {
        name: sql`excluded.name`,
        nameVi: sql`excluded.name_vi`,
        nameEn: sql`excluded.name_en`,
        tags: sql`excluded.tags`,
        minLon: sql`excluded.min_lon`,
        minLat: sql`excluded.min_lat`,
        maxLon: sql`excluded.max_lon`,
        maxLat: sql`excluded.max_lat`,
        centerLon: sql`excluded.center_lon`,
        centerLat: sql`excluded.center_lat`,
        geometryType: sql`excluded.geometry_type`,
        geometry: sql`excluded.geometry`,
        areaApproxKm2: sql`excluded.area_approx_km2`,
        source: sql`excluded.source`,
      },
    });

  const count = batch.length;
  batch.length = 0; // Clear memory
  return count;
}

async function run() {
  const args = process.argv.slice(2);
  const filePath = args[0];
  const batchSize = parseInt(args[1] || '500', 10);

  if (!filePath) {
    console.error('❌ Thiếu đường dẫn file PBF.');
    process.exit(1);
  }

  const resolvedPath = path.resolve(filePath);
  if (!fs.existsSync(resolvedPath)) {
    console.error(`❌ Không tìm thấy file: ${resolvedPath}`);
    process.exit(1);
  }

  const stat = fs.statSync(resolvedPath);
  const fileSizeMb = (stat.size / (1024 * 1024)).toFixed(2);
  const fileName = path.basename(resolvedPath);

  console.log('================================================================');
  console.log(`🚀 BẮT ĐẦU IMPORT FILE OSM PBF VÀO GOOGLE CLOUD SQL POSTGRESQL`);
  console.log(`📁 File: ${fileName} (${fileSizeMb} MB)`);
  console.log(`📦 Batch size: ${batchSize} bản ghi/lần`);
  console.log('================================================================');

  const startTime = Date.now();

  // -------------------------------------------------------------------------
  // PASS 1: Quét nhanh Ways & Relations để thu thập các node ID cần thiết
  // -------------------------------------------------------------------------
  console.log('\n🔍 [Pass 1/2] Đang quét cấu trúc Ways & Relations để lọc Node IDs...');
  const pass1Start = Date.now();

  const neededNodeIds = new Set<number>();
  const pendingWays = new Map<number, { id: number; refs: number[]; tags: OSMTags }>();
  const pendingRelations: Array<{
    id: number;
    tags: OSMTags;
    members: Array<{ type: string; id: number; role?: string }>;
  }> = [];

  let totalWays = 0;
  let totalRels = 0;

  await new Promise<void>((resolve, reject) => {
    const stream1 = fs.createReadStream(resolvedPath);
    const osm1 = parseOSM();

    osm1.on('data', (items: any[]) => {
      for (const item of items) {
        if (item.type === 'way') {
          totalWays++;
          const tags = item.tags || {};
          const isBoundary = tags.boundary === 'administrative' || tags.admin_level !== undefined;
          const isBuilding = tags.building && tags.building !== 'no';
          const isPlace = tags.place !== undefined;
          const isPoi = tags.amenity || tags.tourism || tags.historic || tags.leisure || tags.shop;
          const hasName = Boolean(tags.name || tags['name:vi'] || tags['name:en']);

          if ((isBoundary || isBuilding || isPlace || isPoi || hasName) && item.refs && item.refs.length >= 3) {
            pendingWays.set(item.id, {
              id: item.id,
              refs: item.refs,
              tags,
            });
            for (const ref of item.refs) {
              neededNodeIds.add(ref);
            }
          }
        } else if (item.type === 'relation') {
          totalRels++;
          const tags = item.tags || {};
          const isBoundary = tags.boundary === 'administrative' || tags.admin_level !== undefined;
          const isTypeBoundary = tags.type === 'boundary' || tags.type === 'multipolygon';
          const isBuilding = tags.building && tags.building !== 'no';

          if ((isBoundary || isTypeBoundary || isBuilding) && item.members && item.members.length > 0) {
            pendingRelations.push({
              id: item.id,
              tags,
              members: item.members,
            });
          }
        }
      }
    });

    osm1.on('end', () => resolve());
    osm1.on('error', (err: any) => reject(err));
    stream1.on('error', (err: any) => reject(err));
    stream1.pipe(osm1);
  });

  const pass1Duration = ((Date.now() - pass1Start) / 1000).toFixed(1);
  console.log(`✅ [Pass 1/2] Hoàn tất sau ${pass1Duration}s!`);
  console.log(`   - Tổng Ways: ${totalWays.toLocaleString()} (Chọn lọc: ${pendingWays.size.toLocaleString()})`);
  console.log(`   - Tổng Relations: ${totalRels.toLocaleString()} (Chọn lọc: ${pendingRelations.length.toLocaleString()})`);
  console.log(`   - Số lượng Node IDs cần nạp tọa độ: ${neededNodeIds.size.toLocaleString()}`);

  // -------------------------------------------------------------------------
  // PASS 2: Quét Nodes (lưu tọa độ compact + nạp POI/Địa danh có tên)
  // -------------------------------------------------------------------------
  console.log('\n📥 [Pass 2/2] Đang đọc tọa độ Node & bắt đầu nạp vào PostgreSQL...');
  const pass2Start = Date.now();

  const packedCoords = new Map<number, number>(); // NodeId -> packed number (350MB RAM max)
  let totalSaved = 0;
  const batchBuffer: OSMPlace[] = [];

  let nodesRead = 0;
  let namedNodesSaved = 0;
  let lastLogTime = Date.now();

  // Create stream pipeline with manual pause/resume for safe batch flushing
  await new Promise<void>((resolve, reject) => {
    const stream2 = fs.createReadStream(resolvedPath);
    const osm2 = parseOSM();

    let isFlushing = false;
    const itemQueue: any[][] = [];

    async function processQueue() {
      if (isFlushing) return;
      isFlushing = true;

      while (itemQueue.length > 0) {
        const items = itemQueue.shift()!;
        for (const item of items) {
          if (item.type !== 'node') continue; // In pass 2 we only process nodes; ways/relations are handled next
          nodesRead++;

          if (item.lat !== undefined && item.lon !== undefined) {
            const lat = item.lat;
            const lon = item.lon;

            // Check if node is in Vietnam
            if (lat >= VN_BBOX.minLat && lat <= VN_BBOX.maxLat && lon >= VN_BBOX.minLon && lon <= VN_BBOX.maxLon) {
              // 1. Cache coordinate for ways if needed
              if (neededNodeIds.has(item.id)) {
                packedCoords.set(item.id, packCoord(lon, lat));
              }

              // 2. If node is a named administrative place or POI, insert directly into Postgres
              const tags = item.tags || {};
              const name = tags.name || tags['name:vi'] || tags['name:en'];
              if (
                name &&
                (tags.place ||
                  tags.amenity ||
                  tags.tourism ||
                  tags.historic ||
                  tags.leisure ||
                  tags.shop)
              ) {
                const meta = getAdminMeta(tags);
                batchBuffer.push({
                  id: `node/${item.id}`,
                  osmId: item.id,
                  osmType: 'node',
                  name,
                  nameVi: tags['name:vi'] || name,
                  nameEn: tags['name:en'] || '',
                  adminLevel: meta.adminLevel,
                  adminLevelLabel: meta.adminLevelLabel,
                  priorityRank: meta.priorityRank,
                  placeType: meta.placeType,
                  tags,
                  bbox: {
                    minLon: lon - 0.001,
                    maxLon: lon + 0.001,
                    minLat: lat - 0.001,
                    maxLat: lat + 0.001,
                  },
                  center: [lon, lat],
                  geometryType: 'Point',
                  geometry: { coordinates: [lon, lat] },
                  areaApproxKm2: 0.01,
                });

                if (batchBuffer.length >= batchSize) {
                  stream2.pause();
                  const saved = await flushBatch(batchBuffer, fileName);
                  totalSaved += saved;
                  namedNodesSaved += saved;
                  stream2.resume();
                }
              }
            }
          }
        }

        const now = Date.now();
        if (now - lastLogTime > 4000) {
          lastLogTime = now;
          const memMb = Math.round(process.memoryUsage().heapUsed / 1024 / 1024);
          process.stdout.write(
            `\r⏳ [Nodes] Đã quét: ${nodesRead.toLocaleString()} nodes | Đã lưu: ${totalSaved.toLocaleString()} địa điểm | RAM: ${memMb} MB`
          );
        }
      }

      isFlushing = false;
    }

    osm2.on('data', (items: any[]) => {
      itemQueue.push(items);
      processQueue();
    });

    osm2.on('end', async () => {
      // Wait for any remaining items in queue
      while (itemQueue.length > 0 || isFlushing) {
        await new Promise((r) => setTimeout(r, 50));
      }
      // Flush remaining node POIs
      if (batchBuffer.length > 0) {
        const saved = await flushBatch(batchBuffer, fileName);
        totalSaved += saved;
        namedNodesSaved += saved;
      }
      resolve();
    });

    osm2.on('error', (err: any) => reject(err));
    stream2.on('error', (err: any) => reject(err));
    stream2.pipe(osm2);
  });

  // Free neededNodeIds set from memory now that we have packedCoords
  neededNodeIds.clear();

  console.log(`\n\n✅ Đã nạp thành công ${namedNodesSaved.toLocaleString()} POI và địa danh dạng Node vào Postgres!`);
  console.log(`🔄 Tọa độ sẵn sàng: ${packedCoords.size.toLocaleString()} nodes. Bắt đầu dựng hình học Ways & Ranh giới...`);

  // -------------------------------------------------------------------------
  // PASS 3: Dựng hình học và nạp Ways (Tòa nhà, Phường, Xã, Ranh giới)
  // -------------------------------------------------------------------------
  const wayRings = new Map<number, number[][]>();
  let waysSaved = 0;

  for (const [wayId, way] of pendingWays.entries()) {
    const ring: number[][] = [];
    for (const ref of way.refs) {
      const packed = packedCoords.get(ref);
      if (packed !== undefined) {
        ring.push(unpackCoord(packed));
      }
    }

    if (ring.length < 3) continue;

    // Auto-close polygon ring
    const first = ring[0];
    const last = ring[ring.length - 1];
    if (first[0] !== last[0] || first[1] !== last[1]) {
      ring.push([...first]);
    }

    wayRings.set(wayId, ring);

    const tags = way.tags;
    const isBuilding = tags.building !== undefined && tags.building !== 'no';
    const isBoundary = tags.boundary === 'administrative' || tags.admin_level !== undefined;

    let name = tags.name || tags['name:vi'] || tags['name:en'];
    if (!name && isBuilding) {
      name = formatBuildingName(tags, wayId);
    } else if (!name) {
      name = `Way #${wayId}`;
    }

    const meta = getAdminMeta(tags);
    const bbox = computeBoundingBox([ring]);
    const center: [number, number] = [
      Number(((bbox.minLon + bbox.maxLon) / 2).toFixed(6)),
      Number(((bbox.minLat + bbox.maxLat) / 2).toFixed(6)),
    ];
    const area = calculateApproxAreaKm2([ring]);

    batchBuffer.push({
      id: `way/${wayId}`,
      osmId: wayId,
      osmType: 'way',
      name,
      nameVi: tags['name:vi'] || name,
      nameEn: tags['name:en'] || '',
      adminLevel: meta.adminLevel,
      adminLevelLabel: meta.adminLevelLabel,
      priorityRank: isBuilding ? 0.5 : meta.priorityRank,
      placeType: isBuilding ? 'building' : meta.placeType,
      tags,
      bbox,
      center,
      geometryType: 'Polygon',
      geometry: { coordinates: [ring] },
      areaApproxKm2: area,
    });

    if (batchBuffer.length >= batchSize) {
      const saved = await flushBatch(batchBuffer, fileName);
      totalSaved += saved;
      waysSaved += saved;
      const memMb = Math.round(process.memoryUsage().heapUsed / 1024 / 1024);
      process.stdout.write(`\r💾 [Ways] Đã lưu vào Postgres: ${totalSaved.toLocaleString()} địa điểm | RAM: ${memMb} MB`);
    }
  }

  // Flush remaining ways
  if (batchBuffer.length > 0) {
    const saved = await flushBatch(batchBuffer, fileName);
    totalSaved += saved;
    waysSaved += saved;
  }

  // Free pendingWays map to release memory before relations
  pendingWays.clear();

  console.log(`\n✅ Đã lưu ${waysSaved.toLocaleString()} công trình / ranh giới dạng Way vào Postgres!`);
  console.log(`🔄 Đang lắp ghép ${pendingRelations.length.toLocaleString()} Relations (Cấp tỉnh, quận, huyện, xã, phường)...`);

  // -------------------------------------------------------------------------
  // PASS 4: Lắp ghép và nạp Relations (Đa giác Multipolygon hành chính)
  // -------------------------------------------------------------------------
  let relsSaved = 0;

  for (const rel of pendingRelations) {
    const tags = rel.tags;
    const outerRings: number[][][] = [];

    for (const member of rel.members) {
      if (member.type === 'way') {
        const ring = wayRings.get(member.id);
        if (ring && ring.length >= 3) {
          outerRings.push(ring);
        }
      }
    }

    if (outerRings.length > 0) {
      const name = tags.name || tags['name:vi'] || tags['name:en'] || `Relation #${rel.id}`;
      const meta = getAdminMeta(tags);
      const bbox = computeBoundingBox(outerRings);
      const center: [number, number] = [
        Number(((bbox.minLon + bbox.maxLon) / 2).toFixed(6)),
        Number(((bbox.minLat + bbox.maxLat) / 2).toFixed(6)),
      ];
      const area = calculateApproxAreaKm2(outerRings);

      const isSingle = outerRings.length === 1;
      const geometry: GeoPolygon | GeoMultiPolygon = isSingle
        ? { coordinates: [outerRings[0]] }
        : { coordinates: outerRings.map((r) => [r]) };

      batchBuffer.push({
        id: `relation/${rel.id}`,
        osmId: rel.id,
        osmType: 'relation',
        name,
        nameVi: tags['name:vi'] || name,
        nameEn: tags['name:en'] || '',
        adminLevel: meta.adminLevel,
        adminLevelLabel: meta.adminLevelLabel,
        priorityRank: meta.priorityRank,
        placeType: meta.placeType,
        tags,
        bbox,
        center,
        geometryType: isSingle ? 'Polygon' : 'MultiPolygon',
        geometry,
        areaApproxKm2: area,
      });

      if (batchBuffer.length >= batchSize) {
        const saved = await flushBatch(batchBuffer, fileName);
        totalSaved += saved;
        relsSaved += saved;
      }
    }
  }

  // Flush remaining relations
  if (batchBuffer.length > 0) {
    const saved = await flushBatch(batchBuffer, fileName);
    totalSaved += saved;
    relsSaved += saved;
  }

  // Record audit log
  await db.insert(importLogs).values({
    filename: fileName,
    placesCount: totalSaved,
    fileSize: stat.size,
    status: 'success',
  });

  const durationSec = ((Date.now() - startTime) / 1000).toFixed(1);
  console.log('\n\n================================================================');
  console.log(`🎉 HOÀN TẤT IMPORT DỮ LIỆU TOÀN DIỆN VÀO GOOGLE CLOUD SQL POSTGRESQL!`);
  console.log(`⏱️ Tổng thời gian: ${durationSec}s`);
  console.log(`📍 Tổng số thực thể đã lưu vào PostgreSQL: ${totalSaved.toLocaleString()}`);
  console.log(`   - Địa danh & POI (Nodes): ${namedNodesSaved.toLocaleString()}`);
  console.log(`   - Công trình & Ranh giới (Ways): ${waysSaved.toLocaleString()}`);
  console.log(`   - Ranh giới Hành chính (Relations): ${relsSaved.toLocaleString()}`);
  console.log(`🏠 Dữ liệu đã lưu bền vững tại bảng 'osm_places' trên Cloud SQL.`);
  console.log('================================================================');
  process.exit(0);
}

run().catch((err) => {
  console.error('\n❌ Lỗi trong quá trình import PBF:', err);
  process.exit(1);
});
